src/safeposix/vmmap.rs

```diff
@@ -0,0 +1,587 @@
+        self.entries.get_at_point_mut(page_num)
+    }
+
+    fn last_entry(&self) -> Option<(&Interval<u32>, &VmmapEntry)> {
+        self.entries.last_key_value()
+    }
+
+    fn first_entry(&self) -> Option<(&Interval<u32>, &VmmapEntry)> {
+        self.entries.first_key_value()
+    }
+
+    fn double_ended_iter(&self) -> impl DoubleEndedIterator<Item = (&Interval<u32>, &VmmapEntry)> {
+        self.entries.iter()
+    }
+
+    fn double_ended_iter_mut(
+        &mut self,
+    ) -> impl DoubleEndedIterator<Item = (&Interval<u32>, &mut VmmapEntry)> {
+        self.entries.iter_mut()
+    }
+
+    fn find_page_iter(
+        &self,
+        page_num: u32,
+    ) -> impl DoubleEndedIterator<Item = (&Interval<u32>, &VmmapEntry)> {
+        if let Some(last_entry) = self.last_entry() {
+            self.entries.overlapping(ie(page_num, last_entry.0.end()))
+        } else {
+            // Return an empty iterator if no last_entry
+            self.entries.overlapping(ie(page_num, page_num))
+        }
+    }
+
+    fn find_page_iter_mut(
+        &mut self,
+        page_num: u32,
+    ) -> impl DoubleEndedIterator<Item = (&Interval<u32>, &mut VmmapEntry)> {
+        if let Some(last_entry) = self.last_entry() {
+            self.entries
+                .overlapping_mut(ie(page_num, last_entry.0.end()))
+        } else {
+            // Return an empty iterator if no last_entry
+            self.entries.overlapping_mut(ie(page_num, page_num))
+        }
+    }
+
+    fn find_space(&self, npages: u32) -> Option<Interval<u32>> {
+        let start = self.first_entry();
+        let end = self.last_entry();
+
+        if start == None || end == None {
+            return None;
+        } else {
+            let start_unwrapped = start.unwrap().0.start();
+            let end_unwrapped = end.unwrap().0.end();
+
+            let desired_space = npages + 1; // TODO: check if this is correct
+
+            for gap in self
+                .entries
+                .gaps_trimmed(ie(start_unwrapped, end_unwrapped))
+            {
+                if gap.end() - gap.start() >= desired_space {
+                    return Some(gap);
+                }
+            }
+        }
+
+        None
+    }
+
+    fn find_space_above_hint(&self, npages: u32, hint: u32) -> Option<Interval<u32>> {
+        let start = hint;
+        let end = self.last_entry();
+
+        if end == None {
+            return None;
+        } else {
+            let end_unwrapped = end.unwrap().0.end();
+
+            let desired_space = npages + 1; // TODO: check if this is correct
+
+            for gap in self.entries.gaps_trimmed(ie(start, end_unwrapped)) {
+                if gap.end() - gap.start() >= desired_space {
+                    return Some(gap);
+                }
+            }
+        }
+
+        None
+    }
+
+    fn find_map_space(&self, num_pages: u32, pages_per_map: u32) -> Option<Interval<u32>> {
```

Comment:
It's used to align the start and end of potential memory gaps to specific boundaries, to make sure that the found space is a multiple of pages_per_map and that it is large enough to accommodate the rounded-up num_pages

Reply:
I wonder in which situation we want to find a map space that is multiple of pages_per_map(>1) pages, instead of just multiple of 1 page. Is this something also in nacl? @rennergade

Reply:
pages_per_map probably should be a constant for the min page multiple size we can use for maps in wasm